File: mobile/ios/Utils/ErrorHandler.swift
import Foundation
import UIKit

class ErrorHandler {
    static let shared = ErrorHandler()
    
    private init() {}
    
    // MARK: - Error Handling
    
    func handle(_ error: Error, source: String = #file, line: Int = #line) {
        let errorInfo = ErrorInfo(
            error: error,
            source: source,
            line: line,
            timestamp: Date()
        )
        
        // Log error
        logError(errorInfo)
        
        // Send to analytics
        sendToAnalytics(errorInfo)
        
        // Show user-friendly message
        showUserMessage(for: error)
    }
    
    func handle(_ alamoError: ALADDINError, source: String = #file, line: Int = #line) {
        handle(alamoError as Error, source: source, line: line)
    }
    
    // MARK: - Logging
    
    private func logError(_ errorInfo: ErrorInfo) {
        print("❌ ERROR [\(errorInfo.timestamp)]")
        print("   Source: \(errorInfo.source):\(errorInfo.line)")
        print("   Error: \(errorInfo.error.localizedDescription)")
        
        // В production это было бы отправлено в систему логирования
    }
    
    // MARK: - Analytics
    
    private func sendToAnalytics(_ errorInfo: ErrorInfo) {
        // В реальном приложении здесь была бы отправка в Firebase/Crashlytics
        print("📊 Sending error to analytics...")
    }
    
    // MARK: - User Messages
    
    private func showUserMessage(for error: Error) {
        let message = getUserFriendlyMessage(for: error)
        
        DispatchQueue.main.async {
            if let topViewController = UIApplication.shared.keyWindow?.rootViewController {
                self.showAlert(title: "Ошибка", message: message, on: topViewController)
            }
        }
    }
    
    private func getUserFriendlyMessage(for error: Error) -> String {
        if let alamoError = error as? ALADDINError {
            switch alamoError {
            case .networkError(let networkError):
                return getMessageFor(networkError)
            case .securityError(let securityError):
                return getMessageFor(securityError)
            }
        }
        
        return "Произошла ошибка. Попробуйте позже."
    }
    
    private func getMessageFor(_ networkError: NetworkError) -> String {
        switch networkError {
        case .noConnection:
            return "Нет подключения к интернету"
        case .timeout:
            return "Превышено время ожидания"
        case .serverError(let code):
            return "Ошибка сервера (\(code))"
        case .invalidResponse:
            return "Некорректный ответ сервера"
        case .rateLimited:
            return "Слишком много запросов. Попробуйте позже."
        }
    }
    
    private func getMessageFor(_ securityError: SecurityError) -> String {
        switch securityError {
        case .certificatePinningFailed:
            return "Ошибка безопасности соединения"
        case .invalidCertificate:
            return "Недействительный сертификат"
        case .hostNotTrusted:
            return "Небезопасное соединение"
        }
    }
    
    // MARK: - Alert
    
    private func showAlert(title: String, message: String, on viewController: UIViewController) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default))
        viewController.present(alert, animated: true)
    }
    
    // MARK: - Recovery
    
    func attemptRecovery(from error: Error, completion: @escaping (Bool) -> Void) {
        if let alamoError = error as? ALADDINError {
            switch alamoError {
            case .networkError(.noConnection):
                // Retry with exponential backoff
                retryWithBackoff(completion: completion)
                
            case .networkError(.timeout):
                // Retry immediately
                completion(true)
                
            case .securityError:
                // Cannot recover from security errors
                completion(false)
                
            default:
                completion(false)
            }
        } else {
            completion(false)
        }
    }
    
    private func retryWithBackoff(attempt: Int = 1, completion: @escaping (Bool) -> Void) {
        let delay = min(pow(2.0, Double(attempt)), 30.0) // Max 30 seconds
        
        DispatchQueue.global().asyncAfter(deadline: .now() + delay) {
            // Check if network is available
            completion(true)
        }
    }
}

// MARK: - Error Info

struct ErrorInfo {
    let error: Error
    let source: String
    let line: Int
    let timestamp: Date
}

// MARK: - Result Extension

extension Result {
    func handleError(handler: ErrorHandler = .shared) {
        if case .failure(let error) = self {
            handler.handle(error)
        }
    }
}

File: mobile/android/Utils/ErrorHandler.kt
package com.aladdin.mobile.utils

import android.content.Context
import android.util.Log
import androidx.appcompat.app.AlertDialog
import com.aladdin.mobile.network.ALADDINException
import kotlinx.coroutines.delay
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.math.pow

@Singleton
class ErrorHandler @Inject constructor(
    private val context: Context
) {
    
    // MARK: - Error Handling
    
    fun handle(error: Throwable, source: String = "", line: Int = 0) {
        val errorInfo = ErrorInfo(
            error = error,
            source = source,
            line = line,
            timestamp = System.currentTimeMillis()
        )
        
        // Log error
        logError(errorInfo)
        
        // Send to analytics
        sendToAnalytics(errorInfo)
        
        // Show user-friendly message
        showUserMessage(error)
    }
    
    fun handle(exception: ALADDINException, source: String = "", line: Int = 0) {
        handle(exception as Throwable, source, line)
    }
    
    // MARK: - Logging
    
    private fun logError(errorInfo: ErrorInfo) {
        Log.e(
            "ErrorHandler",
            "ERROR [${java.util.Date(errorInfo.timestamp)}]\n" +
                    "Source: ${errorInfo.source}:${errorInfo.line}\n" +
                    "Error: ${errorInfo.error.message}",
            errorInfo.error
        )
        
        // В production это было бы отправлено в систему логирования
    }
    
    // MARK: - Analytics
    
    private fun sendToAnalytics(errorInfo: ErrorInfo) {
        // В реальном приложении здесь была бы отправка в Firebase/Crashlytics
        Log.i("ErrorHandler", "Sending error to analytics...")
    }
    
    // MARK: - User Messages
    
    private fun showUserMessage(error: Throwable) {
        val message = getUserFriendlyMessage(error)
        
        // В реальном приложении здесь был бы показ диалога
        Log.i("ErrorHandler", "User message: $message")
    }
    
    fun showAlertDialog(title: String, message: String) {
        AlertDialog.Builder(context)
            .setTitle(title)
            .setMessage(message)
            .setPositiveButton("OK", null)
            .show()
    }
    
    private fun getUserFriendlyMessage(error: Throwable): String {
        return when (error) {
            is ALADDINException.NetworkException -> {
                when {
                    error.message?.contains("No internet", ignoreCase = true) == true -> 
                        "Нет подключения к интернету"
                    error.message?.contains("timeout", ignoreCase = true) == true -> 
                        "Превышено время ожидания"
                    error.message?.contains("Server error", ignoreCase = true) == true -> 
                        "Ошибка сервера"
                    else -> "Ошибка сети. Попробуйте позже."
                }
            }
            is ALADDINException.SecurityException -> {
                "Ошибка безопасности соединения"
            }
            is ALADDINException.CertificateException -> {
                "Недействительный сертификат"
            }
            else -> "Произошла ошибка. Попробуйте позже."
        }
    }
    
    // MARK: - Recovery
    
    suspend fun attemptRecovery(error: Throwable): Boolean {
        return when (error) {
            is ALADDINException.NetworkException -> {
                when {
                    error.message?.contains("No internet", ignoreCase = true) == true -> {
                        // Retry with exponential backoff
                        retryWithBackoff()
                    }
                    error.message?.contains("timeout", ignoreCase = true) == true -> {
                        // Retry immediately
                        true
                    }
                    else -> false
                }
            }
            is ALADDINException.SecurityException,
            is ALADDINException.CertificateException -> {
                // Cannot recover from security errors
                false
            }
            else -> false
        }
    }
    
    private suspend fun retryWithBackoff(attempt: Int = 1, maxAttempts: Int = 5): Boolean {
        if (attempt > maxAttempts) {
            return false
        }
        
        val delayMs = min(2.0.pow(attempt.toDouble()), 30.0) * 1000 // Max 30 seconds
        delay(delayMs.toLong())
        
        // Check if network is available
        return isNetworkAvailable()
    }
    
    private fun isNetworkAvailable(): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) 
            as android.net.ConnectivityManager
        val activeNetwork = connectivityManager.activeNetworkInfo
        return activeNetwork?.isConnected == true
    }
    
    // MARK: - Error Types Classification
    
    fun classifyError(error: Throwable): ErrorSeverity {
        return when (error) {
            is ALADDINException.SecurityException,
            is ALADDINException.CertificateException -> ErrorSeverity.CRITICAL
            
            is ALADDINException.NetworkException -> {
                if (error.message?.contains("Server error", ignoreCase = true) == true) {
                    ErrorSeverity.HIGH
                } else {
                    ErrorSeverity.MEDIUM
                }
            }
            
            else -> ErrorSeverity.LOW
        }
    }
}

// MARK: - Data Models

data class ErrorInfo(
    val error: Throwable,
    val source: String,
    val line: Int,
    val timestamp: Long
)

enum class ErrorSeverity {
    LOW, MEDIUM, HIGH, CRITICAL
}

// MARK: - Result Extension

fun <T> Result<T>.handleError(handler: ErrorHandler): Result<T> {
    if (this.isFailure) {
        handler.handle(this.exceptionOrNull()!!)
    }
    return this
}

